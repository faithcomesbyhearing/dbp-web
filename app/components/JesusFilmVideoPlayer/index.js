/**
 *
 * JesusFilmVideoPlayer
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import Hls from 'hls.js';
import VideoControls from '../VideoControls';
import VideoProgressBar from '../VideoProgressBar';
import VideoOverlay from '../VideoOverlay';

class JesusFilmVideoPlayer extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = {
			paused: true,
			volume: 1,
			currentTime: 0,
			bufferLength: 0,
			hlsSupported: true,
		};
		this.videoRef = React.createRef();
	}

	componentDidMount() {
		if (this.videoRef.current) {
			this.initializeHls();
			this.videoRef.current.addEventListener(
				'webkitendfullscreen',
				this.webkitendfullscreen,
			);
			Router.events.on('routeChangeStart', this.handleRouteChange);
		}
		document.addEventListener('keypress', this.pauseWithSpacebar);
	}

	componentWillUnmount() {
		this.cleanupHls();
		if (this.videoRef.current) {
			this.videoRef.current.removeEventListener(
				'webkitendfullscreen',
				this.webkitendfullscreen,
			);
		}
		document.removeEventListener('keypress', this.pauseWithSpacebar);
		Router.events.off('routeChangeStart', this.handleRouteChange);
	}

	webkitendfullscreen = () => {
		this.pauseVideo();
	};

	initializeHls = () => {
		if (Hls.isSupported()) {
			this.hls = new Hls();
			this.hls.on(Hls.Events.MEDIA_ATTACHED, () => {
				this.hls.loadSource(
					`${this.props.hlsStream}?key=${process.env.DBP_API_KEY}&v=4`,
				);
				this.hls.on(Hls.Events.MANIFEST_PARSED, () => {
					// Add user interaction check before playing the video
					this.videoRef.current.addEventListener('click', this.playVideo);
					this.videoRef.current.addEventListener('touchstart', this.playVideo);
				});
			});
			this.hls.attachMedia(this.videoRef.current);
			this.addVideoEventListeners();
		} else if (
			this.videoRef.current.canPlayType('application/vnd.apple.mpegurl')
		) {
			this.videoRef.current.src = `${this.props.hlsStream}?key=${process.env.DBP_API_KEY}&v=4`;
			this.videoRef.current.addEventListener('loadedmetadata', () => {
				// Add user interaction check before playing the video
				this.videoRef.current.addEventListener('click', this.playVideo);
				this.videoRef.current.addEventListener('touchstart', this.playVideo);
			});
			this.addVideoEventListeners();
		} else {
			this.setState({ hlsSupported: false });
		}
	};

	addVideoEventListeners = () => {
		const video = this.hls ? this.hls.media : this.videoRef.current;
		video.addEventListener('timeupdate', this.timeUpdateEventListener);
		video.addEventListener('seeking', this.seekingEventListener);
		video.addEventListener('seeked', this.seekedEventListener);
	};

	cleanupHls = () => {
		if (this.hls) {
			this.hls.destroy();
		}
		if (this.videoRef.current) {
			const video = this.hls ? this.hls.media : this.videoRef.current;

			if (video) {
				video.removeEventListener('timeupdate', this.timeUpdateEventListener);
				video.removeEventListener('seeking', this.seekingEventListener);
				video.removeEventListener('seeked', this.seekedEventListener);
			}
		}
	};

	handleRouteChange = () => {
		this.cleanupHls();
		if (this.videoRef.current) {
			this.pauseVideo();
		}
	};

	togglePlayState = () => {
		const { paused } = this.state;
		if (paused) {
			this.playVideo();
		} else {
			this.pauseVideo();
		}
	};

	timeUpdateEventListener = (e) => {
		this.setState({ currentTime: e.target.currentTime });
	};

	seekingEventListener = (e) => {
		this.setState({ currentTime: e.target.currentTime });
	};

	seekedEventListener = (e) => {
		this.setState({ currentTime: e.target.currentTime });
	};

	playVideo = () => {
		const video = this.hls ? this.hls.media : this.videoRef.current;
		const playPromise = video.play();
		if (playPromise !== undefined) {
			playPromise
				.then(() => {
					this.setState({ paused: false });
				})
				.catch((err) => {
					this.setState({ paused: true });
					if (process.env.NODE_ENV === 'development') {
						console.error('playVideo error', err); // eslint-disable-line no-console
					}
				});
		}
	};

	pauseVideo = () => {
		this.videoRef.current.pause();
		this.setState({ paused: true });
	};

	pauseWithSpacebar = (e) => {
		if (e.key === ' ') {
			this.togglePlayState();
		}
	};

	updateVolume = (volume) => {
		this.videoRef.current.volume = volume;
		this.setState({ volume });
	};

	toggleFullScreen = () => {
		const video = this.videoRef.current;
		if (document.fullscreenElement) {
			document.exitFullscreen();
		} else if (video.requestFullscreen) {
			video.requestFullscreen();
		} else if (video.webkitRequestFullscreen) {
			video.webkitRequestFullscreen();
		} else if (video.mozRequestFullScreen) {
			video.mozRequestFullScreen();
		} else if (video.msRequestFullscreen) {
			video.msRequestFullscreen();
		}
	};

	render() {
		const { volume, paused, currentTime, bufferLength } = this.state;
		const { hasVideo, hlsStream, duration } = this.props;

		if (!hasVideo || !hlsStream) {
			return (
				<div className="video-player-container jesus-film-override">
					<h2>
						Something went wrong. Please refresh the page or go back to try
						again!
					</h2>
				</div>
			);
		}

		return (
			<div className="video-player-container jesus-film-override">
				<div className="video-player">
					<VideoOverlay
						paused={paused}
						playFunction={this.playVideo}
						pauseFunction={this.pauseVideo}
						togglePlayState={this.togglePlayState}
						isJesusFilm
					/>
					<video ref={this.videoRef} onClick={this.togglePlayState} />
					<VideoProgressBar
						paused={paused}
						currentTime={currentTime}
						duration={duration}
						setCurrentTime={this.setCurrentTime}
						bufferLength={bufferLength}
					/>
					<VideoControls
						paused={paused}
						pauseVideo={this.pauseVideo}
						toggleFullScreen={this.toggleFullScreen}
						updateVolume={this.updateVolume}
						volume={volume}
					/>
				</div>
			</div>
		);
	}
}

JesusFilmVideoPlayer.propTypes = {
	hlsStream: PropTypes.string.isRequired,
	duration: PropTypes.number.isRequired,
	hasVideo: PropTypes.bool.isRequired,
};

export default JesusFilmVideoPlayer;
