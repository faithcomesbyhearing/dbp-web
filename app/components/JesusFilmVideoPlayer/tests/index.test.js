import React from 'react';
import renderer from 'react-test-renderer';

import JesusFilmVideoPlayer from '..';

const props = {
  hlsStream:
    'https://api-dev.dbp4.org/arclight/jesus-film?key=${process.env.DBP_API_KEY}&v=4&arclight_id=23156',
  duration: 5789,
  hasVideo: true,
};

describe('<JesusFilmVideoPlayer /> component tests', () => {
  it('Should match snapshot with default props', () => {
    const tree = renderer.create(<JesusFilmVideoPlayer {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
  it('Should match snapshot with no hls stream', () => {
    const tree = renderer
      .create(<JesusFilmVideoPlayer {...props} hlsStream={''} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
