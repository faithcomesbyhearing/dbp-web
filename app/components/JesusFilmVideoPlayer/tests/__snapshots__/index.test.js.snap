// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JesusFilmVideoPlayer /> component tests Should match snapshot with default props 1`] = `
<div
  className="video-player-container jesus-film-override"
>
  <div
    className="video-player"
  >
    <div
      className="play-video-container show-control-icon"
      onClick={[Function]}
    >
      <div
        className="control-button-container"
        onClick={[Function]}
      >
        <span
          className="play-video-title"
        >
          Jesus Film
        </span>
        <svg
          className="play-video"
          fill="#fff"
        >
          <use
            xlinkHref="/svglist.svg#play_video"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          />
        </svg>
      </div>
    </div>
    <video
      onClick={[Function]}
    />
    <div
      className="progress-slider hide-progress-slider"
      data-value-cur="00:00"
      data-value-dur="01:36:29"
    >
      <div
        className="progress-bar-buffer"
        style={
          {
            "width": "0%",
          }
        }
      />
      <div
        className="rc-slider rc-slider-horizontal"
        onMouseDown={[Function]}
      >
        <div
          className="rc-slider-rail"
          style={
            {
              "backgroundColor": "rgba(199, 199, 204, 0.95)",
              "cursor": "pointer",
              "height": "2px",
            }
          }
        />
        <div
          className="rc-slider-track"
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            {
              "backgroundColor": "rgba(98, 177, 130, 1)",
              "cursor": "pointer",
              "height": "2px",
              "left": "0%",
              "width": "0%",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          aria-disabled={false}
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={0}
          className="rc-slider-handle"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          style={
            {
              "backgroundColor": "rgba(98, 177, 130, 1)",
              "border": "2px solid rgba(98, 177, 130, 1)",
              "height": "10px",
              "left": "0%",
              "top": "6px",
              "transform": "translateX(-50%)",
              "width": "10px",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
    <div
      className="controls hide-controls"
    >
      <div
        className="left-controls"
      >
        <div
          className="video-volume-container"
          onClick={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="video-volume-slider-container"
          >
            <div
              className="video-slider-container"
            >
              <div
                className="rc-slider slider rc-slider-vertical"
                onMouseDown={[Function]}
              >
                <div
                  className="rc-slider-rail"
                  style={
                    {
                      "backgroundColor": "#000",
                      "width": "2px",
                    }
                  }
                />
                <div
                  className="rc-slider-track"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                  style={
                    {
                      "backgroundColor": "rgba(98, 177, 130, 1)",
                      "bottom": "0%",
                      "height": "100%",
                      "width": "2px",
                    }
                  }
                />
                <div
                  className="rc-slider-step"
                />
                <div
                  aria-disabled={false}
                  aria-valuemax={100}
                  aria-valuemin={0}
                  aria-valuenow={100}
                  className="rc-slider-handle"
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                  role="slider"
                  style={
                    {
                      "backgroundColor": "rgba(98, 177, 130, 1)",
                      "borderColor": "rgba(98, 177, 130, 1)",
                      "bottom": "100%",
                      "height": "12.5px",
                      "left": "5px",
                      "transform": "translateY(50%)",
                      "width": "12.5px",
                    }
                  }
                  tabIndex={0}
                />
              </div>
            </div>
          </div>
          <svg
            className="icon"
            fill="#fff"
          >
            <use
              xlinkHref="/svglist.svg#volume_max"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
        <div
          onClick={[Function]}
        >
          <svg
            fill="#fff"
          >
            <use
              xlinkHref="/svglist.svg#pause"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
      </div>
      <div
        className="right-controls"
      >
        <div>
          <svg
            className="video-elipsis"
            fill="#fff"
          >
            <use
              xlinkHref="/svglist.svg#elipsis"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
        <div
          onClick={[Function]}
        >
          <svg
            className="video-fullscreen"
            fill="#fff"
          >
            <use
              xlinkHref="/svglist.svg#fullscreen"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<JesusFilmVideoPlayer /> component tests Should match snapshot with no hls stream 1`] = `
<div
  className="video-player-container jesus-film-override"
>
  <h2>
    Something went wrong. Please refresh the page or go back to try again!
  </h2>
</div>
`;
